src/sentry/debug_files/artifact_bundles.py

```diff
@@ -0,0 +1,206 @@
+        if set_artifact_bundle_being_indexed_if_null(
+            organization_id=organization_id, artifact_bundle_id=artifact_bundle.id
+        ):
+            bundles_to_index.append(artifact_bundle)
+        else:
+            # A different asynchronous job is taking care of this bundle.
+            metrics.incr("artifact_bundle_indexing.bundle_already_being_indexed")
+
+    artifact_bundles = bundles_to_index
+    if not artifact_bundles:
+        return
+
+    # Second: we merge in-memory all the files of the bundles to index, in order to obtain the minimum set of changes
+    # that we can apply on the database (e.g., two bundles with identical files will end up with just the newest files
+    # actually indexed in the database).
+    files_to_index: Dict[str, ArtifactBundle] = {}
+
+    for artifact_bundle in artifact_bundles:
+        archive = ArtifactBundleArchive(artifact_bundle.file.getfile(), build_memory_map=False)
+        try:
+            for info in archive.get_files().values():
+                url = info.get("url")
+                if url is None:
+                    continue
+
+                indexed = files_to_index.get(url)
+                # In order to obtain a deterministic total order of bundles, we want to first compare their date and
+                # then their id, just to discriminate in case of equal date.
+                bundle_ordering = (artifact_bundle.date_last_modified, artifact_bundle.id)
+                if not indexed or (indexed.date_last_modified, indexed.id) < bundle_ordering:
+                    files_to_index[url] = artifact_bundle
+        finally:
+            archive.close()
+
+    # Third: we commit the intended changes to the database, carefully merging with the state that already exists
+    # in the database. We do so grouping by bundle, using one atomic transaction for the indexing and setting of
+    # the indexed state as well.
+    urls_by_bundle: DefaultDict[ArtifactBundle, Set[str]] = defaultdict(set)
+    for url, bundle in files_to_index.items():
+        urls_by_bundle[bundle].add(url)
+
+    # The naming here is a bit confusing, it is the time we touched the index, and is used for partitioning the index
+    # table and auto-expiration.
+    date_added = datetime.now()
+
+    # Finally: we have to loop over all bundles, since we have to mark all the ones we inspected as `was_indexed`,
+    # irrespectively if we have urls to add for that bundle.
+    for artifact_bundle in artifact_bundles:
+        urls = urls_by_bundle.get(artifact_bundle, set())
+        # We want to start a transaction for each bundle, so that in case of failures we keep consistency at the
+        # bundle level, and we also have to retry only the failed bundle in the future and not all the bundles.
+        with atomic_transaction(
+            using=(
+                router.db_for_write(ArtifactBundleIndex),
+                router.db_for_write(ArtifactBundle),
+            )
+        ):
```

Comment:
This style of transaction opening (with multiple models set) will create either create multiple savepoints on a single connection, or create separate transactions on different connections, but there is no atomicity promise/guarantee that transactions on multiple connections will succeed/fail together.

Personally I would avoid this helper as it provides an interface that either adds overhead or promises things that cannot be delivered.